examples/timer_functional/src/main.rs

```diff
@@ -0,0 +1,170 @@
+use std::rc::Rc;
+
+use gloo::timers::callback::{Interval, Timeout};
+use yew::prelude::*;
+
+fn get_current_time() -> String {
+    let date = js_sys::Date::new_0();
+    String::from(date.to_locale_time_string("en-US"))
+}
+
+enum TimerAction {
+    Add(&'static str),
+    Cancel,
+    SetInterval(Interval),
+    SetTimeout(Timeout),
+    TimeoutDone,
+}
+
+#[derive(Clone, Debug)]
+struct TimerState {
+    messages: Vec<&'static str>,
+    interval_handle: Option<Rc<Interval>>,
+    timeout_handle: Option<Rc<Timeout>>,
+}
+
+impl PartialEq for TimerState {
+    fn eq(&self, other: &Self) -> bool {
+        self.messages == other.messages
+            && self.interval_handle.is_some() == other.interval_handle.is_some()
+    }
+}
+
+impl Reducible for TimerState {
+    type Action = TimerAction;
+
+    fn reduce(self: Rc<Self>, action: TimerAction) -> Rc<Self> {
+        match action {
+            TimerAction::Add(message) => {
+                let mut messages = self.messages.clone();
+                messages.push(message);
+                Rc::new(TimerState {
+                    messages,
+                    interval_handle: self.interval_handle.clone(),
+                    timeout_handle: self.timeout_handle.clone(),
+                })
+            }
+            TimerAction::SetInterval(t) => Rc::new(TimerState {
+                messages: vec!["Interval started!"],
+                interval_handle: Some(Rc::from(t)),
+                timeout_handle: self.timeout_handle.clone(),
+            }),
+            TimerAction::SetTimeout(t) => Rc::new(TimerState {
+                messages: vec!["Timer started!!"],
+                interval_handle: self.interval_handle.clone(),
+                timeout_handle: Some(Rc::from(t)),
+            }),
+            TimerAction::TimeoutDone => {
+                let mut messages = self.messages.clone();
+                messages.push("Done!");
+                Rc::new(TimerState {
+                    messages,
+                    interval_handle: self.interval_handle.clone(),
+                    timeout_handle: None,
+                })
+            }
+            TimerAction::Cancel => {
+                let mut messages = self.messages.clone();
+                messages.push("Canceled!");
+                Rc::new(TimerState {
+                    messages,
+                    interval_handle: None,
+                    timeout_handle: None,
+                })
+            }
+        }
+    }
+}
+
+#[function_component(Clock)]
+fn clock() -> Html {
+    let time = use_state(get_current_time);
+
+    {
+        let time = time.clone();
+        use_effect_with_deps(
```

Comment:
I think we just updated the function signature of some hooks in #2861. Could you please merge master and update the hook?

Reply:
Done